sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -205,30 +206,145 @@ class Analyzer(
+      }
+
+      plan match {
+        case SubqueryAlias(name, u: Union) if recursiveTableName.isDefined =>
+          def combineUnions(union: Union): Seq[LogicalPlan] = union.children.flatMap {
+            case u: Union => combineUnions(u)
+            case o => Seq(o)
           }
+
+          val substitutedTerms = combineUnions(u).map(substitute(_))
+          val (anchorTerms, recursiveTerms) = substitutedTerms.partition(!_._2)
+
+          if (!recursiveTerms.isEmpty) {
+            if (anchorTerms.isEmpty) {
+              throw new AnalysisException("There should be at least 1 anchor term defined in a " +
+                s"recursive query $name")
+            }
+
+            val recursiveTermPlans = recursiveTerms.map(_._1)
+
+            def traversePlanAndCheck(
+                plan: LogicalPlan,
+                isRecursiveReferenceAllowed: Boolean = true): Boolean = plan match {
+              case UnresolvedRecursiveReference(name) =>
+                if (!isRecursiveReferenceAllowed) {
+                  throw new AnalysisException(s"Wrong usage of recursive reference ${name}")
+                }
+                true
+              case Join(left, right, Inner, _, _) =>
+                val l = traversePlanAndCheck(left, isRecursiveReferenceAllowed)
+                val r = traversePlanAndCheck(right, isRecursiveReferenceAllowed)
+                if (l && r) {
+                  throw new AnalysisException("Recursive reference can't be used in on both " +
+                    "side of an inner join")
+                }
+                l || r
+              case Join(left, right, LeftOuter, _, _) =>
+                traversePlanAndCheck(left, isRecursiveReferenceAllowed) ||
+                  traversePlanAndCheck(right, false)
+              case Join(left, right, RightOuter, _, _) =>
+                traversePlanAndCheck(left, false) ||
+                  traversePlanAndCheck(right, isRecursiveReferenceAllowed)
+              case Join(left, right, FullOuter, _, _) =>
+                traversePlanAndCheck(left, false) || traversePlanAndCheck(right, false)
+              case Aggregate(_, _, child) => traversePlanAndCheck(child, false)
+              case Distinct(child) => traversePlanAndCheck(child, false)
+              case o =>
+                o.children.map(traversePlanAndCheck(_, isRecursiveReferenceAllowed)).contains(true)
+            }
+
+            recursiveTermPlans.foreach(traversePlanAndCheck(_))
+
+            RecursiveTable(
+              recursiveTableName.get,
+              SubqueryAlias(name, Union(anchorTerms.map(_._1))),
+              Union(recursiveTermPlans))
+          } else {
+            SubqueryAlias(name, Union(substitutedTerms.map(_._1)))
+          }
+
+        case _ =>
+          val (substitutedPlan, recursiveReferenceFound) = substitute(plan)
+
+          if (recursiveReferenceFound) {
+            throw new AnalysisException("Wrong usage of recursive reference " +
```

Comment:
Wrong usage sounds a bit unclear. Is it possible to have error message explaining why it is wrong?